File: command/src/main/java/com/nearsoft/workshop/command/java/Invoker.java
package com.nearsoft.workshop.command.java;

import java.util.ArrayList;
import java.util.List;

public class Invoker {

    private List<Command> executedCommands = new ArrayList<>();

    public void executeCommand(Command purchase) {
        purchase.execute();
        executedCommands.add(purchase);
    }

    public void replay(){
        for (Command command : executedCommands) {
            command.execute();
        }
    }
}

File: command/src/main/scala/com/nearsoft/workshop/command/scala/Invoker.scala
package com.nearsoft.workshop.command.scala

class Invoker {
  var commands: Vector[() => Unit] = Vector()

  def executeCommand(command: () => Unit) = {
    commands = commands :+ command
    command()
  }

  def replayCommands = {
    commands.foreach(_.apply())
  }
}
